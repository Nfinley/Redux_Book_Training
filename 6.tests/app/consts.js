export const ADD_RECIPE    = 'ADD_RECIPE';
export const DEL_RECIPE    = 'DEL_RECIPE';
export const SET_RECIPES   = 'SET_RECIPES';
export const SET_RECIPE    = 'SET_RECIPE';
export const API           = 'API';
export const API_STARTED   = 'API_STARTED';
export const API_FINISHED  = 'API_FINISHED';
export const API_ERROR     = 'API_ERROR';
export const FETCH_ERROR   = 'FETCH_ERROR';