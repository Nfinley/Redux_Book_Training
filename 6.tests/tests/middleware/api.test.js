import apiMiddleware from 'middleware/api';
import { mockFetch, mockFetchError } from 'test-utils';
import { API_STARTED, API_FINISHED, API, API_ERROR } from 'consts';

const data = { title: 'hello' };

const setData = data => ({
  type: 'SET_DATA',
  payload: data
});

const apiAction = () => ({
  type: API,
  payload: {
    success: setData,
    url: 'fake.json'
  }
});

describe("api middleware", () => {
  let next, dispatch, middleware, dispatchCalls, nextCalls, getState, fetch;

  const getStateMock = () => getState ? getState() : {};

  beforeEach(() => {
    getState      = null;
    next          = jest.fn();
    dispatch      = jest.fn();
    dispatchCalls = dispatch.mock.calls;
    nextCalls     = next.mock.calls;
    middleware    = apiMiddleware({ dispatch, getState: getStateMock })(next);
  });

  it('should ignore non API actions', () => {
    const sampleAction = { type: 'SAMPLE_ACTION' };

    middleware(sampleAction);

    expect(dispatchCalls.length).toBe(0);
    expect(nextCalls).toEqual([[sampleAction]]);
  });

  it('should dispatch API_STARTED', () => {
    middleware(apiAction());
    expect(dispatchCalls[0]).toEqual([{ type: API_STARTED }]);
  });

  describe('success', () => {
    beforeEach(() => mockFetch('fake.json', 200, JSON.stringify(data)));

    it('should dispatch API_FINISHED', () =>
      middleware(apiAction()).then(() =>
        expect(dispatchCalls[2]).toMatchSnapshot()));

    it('should dispatch SET_DATA', () =>
      middleware(apiAction()).then(() =>
        expect(dispatchCalls[1]).toEqual([setData(data)])));
  });

  describe('error', () => {
    beforeEach(() => mockFetchError('fake.json', 404, 'Not found'));

    it('should NOT dispatch API_FINISHED', () =>

      middleware(apiAction()).then(() =>
        expect(dispatchCalls[1][0].type).not.toBe(API_FINISHED)));

    it('should dispatch error', () =>
      middleware(apiAction()).then(() =>
        expect(dispatchCalls[1]).toMatchSnapshot()));
  });

  describe('headers', () => {
    beforeEach(() => fetch = mockFetch('fake.json', 200, JSON.stringify(data)));

    it('should set access token if in state', () => {
      getState = () => ({ accessToken: 'xx11' });

      return middleware(apiAction()).then(() => {
        expect(fetch.mock.calls[0][1].headers.get('Access-Token')).toEqual('xx11')
      });
    });

    it('should NOT set access token if NOT in state', () => {
      getState = () => ({ });

      return middleware(apiAction()).then(() => {
        expect(fetch.mock.calls[0][1].headers.get('Access-Token')).toEqual(null)
      });
    })
  });
});
