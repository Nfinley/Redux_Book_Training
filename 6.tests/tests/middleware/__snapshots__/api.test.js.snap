exports[`api middleware error should dispatch error 1`] = `
Array [
  Object {
    "error": true,
    "payload": [Error: [object Object]],
    "type": "API_ERROR",
  },
]
`;

exports[`api middleware success should dispatch API_FINISHED 1`] = `
Array [
  Object {
    "type": "API_FINISHED",
  },
]
`;
